helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -1261,31 +1257,75 @@ private void processDataOrChildChange(WatchedEvent event, long notificationTime)
 
     if (event.getType() == EventType.NodeDataChanged || event.getType() == EventType.NodeDeleted
         || event.getType() == EventType.NodeCreated) {
+      boolean pathExists = event.getType() != EventType.NodeDeleted;
       Set<IZkDataListenerEntry> listeners = _dataListener.get(path);
       if (listeners != null && !listeners.isEmpty()) {
-        fireDataChangedEvents(event.getPath(), listeners, OptionalLong.of(notificationTime));
+        fireDataChangedEvents(event.getPath(), listeners, notificationTime, pathExists);
+      }
+    }
+  }
+
+  /**
+   * The method is called when state change occurs and we need to handle existing path data handlers
+   */
+  private void fireDataChangedEvents(final String path, Set<IZkDataListenerEntry> listeners) {
```

Comment:
I do see some duplicated code between these three methods. Can you combine the logic to optimize the code structure.